src/protocol/libp2p/kademlia/mod.rs

```diff
@@ -758,6 +766,52 @@ impl Kademlia {
 
                 Ok(())
             }
+            QueryAction::AddProviderToFoundNodes { provider, peers } => {
+                tracing::trace!(
+                    target: LOG_TARGET,
+                    provided_key = ?provider.key,
+                    num_peers = ?peers.len(),
+                    "add provider record to found peers",
+                );
+
+                let provided_key = provider.key.clone();
+                let message = KademliaMessage::add_provider(provider);
+                let peer_action = PeerAction::SendAddProvider(message);
+
+                for peer in peers {
+                    match self.service.open_substream(peer.peer) {
+                        Ok(substream_id) => {
+                            self.pending_substreams.insert(substream_id, peer.peer);
+                            self.peers
+                                .entry(peer.peer)
+                                .or_default()
+                                .pending_actions
+                                .insert(substream_id, peer_action.clone());
+                        }
+                        Err(_) => match self.service.dial(&peer.peer) {
+                            Ok(_) => match self.pending_dials.entry(peer.peer) {
+                                Entry::Occupied(entry) => {
+                                    entry.into_mut().push(peer_action.clone());
+                                }
```

Comment:
Could we use `open_substream_or_dial` instead? That wrapper also handles retries wrt Error::AlreadyConnected and races between opening substreams and dialing

Reply:
Done